### PR TITLE
Make send/reject injectable

### PR DESCRIPTION
**Pull request by bajtos**
As I started working on `app.sequence((seq, res, res) => {...})`, I realised that we really need to refactor `sendResponse`/`sendError` into injectable `send`/`reject`.  That's what this pull request presents.

cc @bajtos @raymondfeng @ritch @superkhau

### CODE REVIEW

**Comment by kjdelisle** [comment]
What I see looks good, but I don't see any tests.

**Comment by bajtos** [comment]
@kjdelisle good call. 

It turned out that our testlab infrastructure is not good enough yet to make it easy to write the tests the way I'd like them to be. I improved testlab helpers while adding the new tests.

I added two kinds of tests:
 - unit-tests for `RejectProvider`, verifying how `reject` works in insolation
 - acceptance tests verifying that applications can bind their custom `send` and `reject` implementations, while using the default `Sequence`.

@raymondfeng @ritch @superkhau @kjdelisle PTAL again.

File: packages/core/test/integration/http-handler.integration.ts [hunk 1]
```diff
@@ -402,7 +410,14 @@ describe('HttpHandler', () => {
   }
 
   function givenClient() {
-    client = createClientForHandler(handler.handleRequest);
+    client = createClientForHandler((req, res) => {
+      handler.handleRequest(req, res).catch(err => {
+        console.error('Request failed.', err.stack);
```
Comment (bajtos) [inline packages/core/test/integration/http-handler.integration.ts]: This catch block is needed to prevent unhandled rejection warnings in the test output.

File: packages/core/test/acceptance/sequence/sequence.acceptance.ts [hunk 2]
```diff
@@ -68,6 +70,28 @@ describe('Sequence - ', () => {
...
+    };
+    app.bind('send').to(send);
+
+    const client = await whenIMakeRequestTo(app);
+    await client.get('/name').expect('CUSTOM FORMAT: SequenceApp');
+  });
+
+  it('user-defined Reject', async () => {
+    const reject: Reject = (response, request, error) => {
+      response.statusCode = 418; // I'm a teapot
```
Comment (kjdelisle) [inline packages/core/test/acceptance/sequence/sequence.acceptance.ts]: Shouldn't this technically be an error, since it's a 4XX code? :P

File: packages/testlab/src/testlab.ts [hunk 3]
```diff
@@ -6,14 +6,17 @@
 /// <reference path="../should-as-function.d.ts" />
 
 const shouldAsFunction: Internal = require('should/as-function');
+require('should-sinon');
```
Comment (kjdelisle) [inline packages/testlab/src/testlab.ts]: Unless I'm mistaken, you should be able to write this as `import 'should-sinon';` and it will do all the same global importing.

File: packages/core/test/acceptance/sequence/sequence.acceptance.ts (hunk 2 above)
Comment (bajtos) [inline packages/core/test/acceptance/sequence/sequence.acceptance.ts]: I am not sure that I understand. What should be technically an error?

`reject` is called when request handling failed, therefore the HTTP status code should be one of 4xx or 5xx.

File: packages/testlab/src/testlab.ts (hunk 3 above)
Comment (bajtos) [inline packages/testlab/src/testlab.ts]: Yep, `import` works well. I'll make the change as you are proposing.

File: packages/core/src/application.ts [hunk 4]
```diff
@@ -50,6 +52,8 @@ export class Application extends Context {
       this._handleHttpRequest(req, res);
 
     this.bind('logError').to(this._logError.bind(this));
+    this.bind('send').to(writeResultToResponse);
```
Comment (raymondfeng) [inline packages/core/src/application.ts]: I start to wonder if we should use a prefix for such actions, for example:
`actions.send` or even `sequence.actions.send`.

Comment (bajtos) [inline packages/core/src/application.ts]: Makes sense, I always had a bit of a weird feeling about using "global" non-descriptive names like "send"/"reject".

I like `sequence.actions.send` 👍

**Comment by bajtos** [comment]
@raymondfeng I have added "sequence.actions." prefix, PTAL.

File: packages/core/src/sequence.ts [hunk 5]
```diff
@@ -45,7 +47,8 @@ export class Sequence {
   constructor(
     @inject('findRoute') protected findRoute: FindRoute,
```
Comment (raymondfeng) [inline packages/core/src/sequence.ts]: Do we want to rename the keys as `sequence.actions.*` here or in a separate PR?

File: packages/core/test/acceptance/sequence/sequence.acceptance.ts (hunk 2 above)
Comment (kjdelisle) [inline packages/core/test/acceptance/sequence/sequence.acceptance.ts]: I just meant that you'd want the response body to contain some sort of message, or something to that effect, rather than just 418, but it doesn't *actually* matter since it's just a test of the ability to define your rejection handler.

Comment (kjdelisle) [inline packages/core/test/acceptance/sequence/sequence.acceptance.ts]: And also because it doesn't need a message to be an error. Aside from that, do we intend to change how we present errors in response bodies as compared to loopback@3.x?

File: packages/core/src/sequence.ts (hunk 5 above)
Comment (bajtos) [inline packages/core/src/sequence.ts]: Let's do that in a separate PR.
